mesonbuild/interpreter/interpreter.py

```diff
@@ -722,9 +722,10 @@ def func_declare_dependency(self, node: mparser.BaseNode, args: T.List[TYPE_var]
             except ValueError:
                 continue
             else:
-                if not self.is_subproject() and srcdir / self.subproject_dir in p.parents:
-                    continue
-                if p.is_absolute() and p.is_dir() and srcdir / self.root_subdir in [p] + list(Path(os.path.abspath(p)).parents):
+                if p.is_absolute() \
+                        and (self.is_subproject() or srcdir / self.subproject_dir not in p.parents) \
+                        and srcdir / self.root_subdir in [p] + list(Path(os.path.abspath(p)).parents) \
+                        and p.is_dir():
```

Comment:
This feels complex and suboptimal...

- you compute p parents two times...
- p.is_dir() should be done before (just after the p.is_absolute() ) to short-circuit the test 
- why os.path.abspath(p) instead of p.resolve() ?
- string comparison or relative path computation would probably be faster than iterating over parent paths
- I wonder if an helper function would make it easier to read

Reply:
Computing `p.dir()` last is actually *the* fix for the bug. :)  I did move `p.is_absolute()` first, before the first expensive `in p.parents`, could already be an improvement, but the code was not optimized before and I didn't want to change the path manipulation logic in any way.

This explains why I kept the preexisting weird mix of pathlib and os.path, for example the two calls to `parents` are on subtly different paths. However, if #14110 goes in, I want to get rid of all occurrences of `in x.parents` which is quadratic (and with a stupidly large constant in front too).

Reply:
I see... Then I think it worth adding a comment telling why is_dir must be last...